### PR TITLE
Consider hub_dir alongside TORCH_HOME env variable for storing hub models

### LINKED ISSUE #31944: torch.hub.set_dir() is ignored for downloaded models
## 🐛 Bug

I use torch.hub.set_dir(localdir) in order to avoid having the model files hidden in `.cache/torch/pytorch_fairseq/3f864e15bb396f062dd37494309dbc4238416edd1f8e...`, but only `pytorch_fairseq_master` goes there, the model itself (`model.pt`, `dict.txt`) are still in `.cache/torch/pytorch_fairseq/3f864e1...`

## To Reproduce

```python
import torch
torch.hub.set_dir("/opt/models/xlmr.large")
model = torch.hub.load('pytorch/fairseq', 'xlmr.large')
```

```bash
$> ls /opt/models/xlmr.large
pytorch_fairseq_master
```

## Expected behaviour

According to the documentation of `torch.hub.set_dir()` I expected the following in `/opt/models/xlmr.large`:

- pytorch_fairseq_master/
- dict.txt
- model.pt
- sentencepiece.bpe.model

But the latter three reside in `.cache/torch/pytorch_fairseq/3f864e15bb396f...`

## Environment
 - PyTorch Version (e.g., 1.0): 1.3.1
 - OS (e.g., Linux): Linux (Ubuntu 18.04.3 LTS)
 - GCC: 7.4.0
 - CMake: version 3.10.2

 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Python version: 3.7
 - GPU
    - GPU 0: GeForce GTX 1080 Ti
    - GPU 1: GeForce GTX 1080 Ti

cc @ezyang @gchanan @zou3519 @ailzhang

### PR DESCRIPTION
**Pull request by andfoy**
Fixes #31944

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 62328c60:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=94ddc2c462bd7c79426057d67b03d4a397db0dd1&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 94ddc2c4 since Jan 29
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=62328c602ff0125119d7ce4bd4fd764009356138).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_linux_backward_compatibility_check_test](https://dr.pytorch.org/build-details.html?build_id=93508331) from Jan 29



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Review by rgommers — COMMENTED** [review]
@andfoy thanks. This looks correct, but can you please add to the PR description what you tested?

**Comment by ezyang** [comment]
deferring review to @ailzhang

**Review by ailzhang — CHANGES_REQUESTED** [review]
Thanks for fixing it @andfoy ! 
Yea agree with @rgommers that adding a test would be great! 
It could be as simple as calling `hub.set_dir` and then check the output of `torch.hub._get_torch.home()`. (Yea this is not supposed to be a public API, but good to keep as fairseq is using it ;)

**Review by ailzhang — CHANGES_REQUESTED** [review]
@andfoy Hey sorry, I should have mentioned that we have `TestHub` in `test_utils.py`. Could you move the test over there?

**Comment by andfoy** [comment]
> @andfoy Hey sorry, I should have mentioned that we have TestHub in test_utils.py. Could you move the test over there?

No worries, I'm moving it right now!

File: test/run_test.py [hunk 1]
```diff
@@ -64,7 +64,7 @@
     'test_type_promotion',
     'test_jit_disabled',
     'test_function_schema',
-    'test_overrides',
```
Comment (ailzhang) [inline test/run_test.py]: nit: can we revert this line as well ;) ?

Comment (andfoy) [reply test/run_test.py]: Shall I keep the trailling comma?

**Review by ailzhang — APPROVED** [review]
Thanks!!

**Comment by ailzhang** [comment]
@andfoy Ah it failed py2 test since there's no TemporaryDirectory in tempfile. Given that we already dropped py2 support publicly, I think simply adding 
```
from torch._siz import PY2
@unittest.skipIf(PY2, "Requires python 3")
```
will work. Would you mind updating it? Thanks a lot!

**Pull request closed without merging**